panther_driver/src/panther_can.py

```diff
@@ -0,0 +1,169 @@
+                    try:
+                        # division by 10 is needed according to documentation
+                        motor_controller.wheel_curr[i] = motor_controller.can_node.sdo['Qry_MOTAMPS'][wheel].raw / 10.0
+                    except canopen.SdoCommunicationError:
+                        rospy.logwarn(
+                            f'[{rospy.get_name()}] PantherCAN: SdoCommunicationError ' 
+                            'occurred while reading motor current'
+                        )
+                        self._error_handle()
+                    yield motor_controller.wheel_curr[i]
+
+    def get_battery_data(self): 
+        with self._lock:
+            for motor_controller in self._motor_controllers:
+                try:
+                    # division by 10 is needed according to documentation
+                    motor_controller.battery_data[0] = float(motor_controller.can_node.sdo['Qry_VOLTS'][VOLT_CHANNEL].raw) / 10.0
+                    motor_controller.battery_data[1] = float(motor_controller.can_node.sdo['Qry_BATAMPS'][AMP_CHANNEL].raw) / 10.0
+                except canopen.SdoCommunicationError:
+                    rospy.logwarn(
+                        f'[{rospy.get_name()}] PantherCAN: SdoCommunicationError ' 
+                        'occurred while reading battery data'
+                    )       
+                    self._error_handle()
+                
+                for i in range(2):
+                    yield motor_controller.battery_data[i]
+
+    def read_fault_flags(self):
+        with self._lock:
+            for motor_controller in self._motor_controllers:
+                try:
+                    motor_controller.fault_flags = motor_controller.can_node.sdo['Qry_FLTFLAG'].raw
+                except canopen.SdoCommunicationError:
+                    rospy.logwarn(
+                        f'[{rospy.get_name()}] PantherCAN: SdoCommunicationError occurred while reading fault flags'
+                    )
+                    self._error_handle()
+                yield motor_controller.fault_flags
+
+    def read_runtime_stat_flag(self):
+        with self._lock:
+            for motor_controller in self._motor_controllers:
+                for i, wheel in enumerate([LEFT_WHEEL, RIGHT_WHEEL]):
+                    try:
+                        motor_controller.runtime_stat_flag[i] = int.from_bytes(motor_controller.can_node.sdo['Qry_MOTFLAGS'][wheel].data, 'little') 
+                    except canopen.SdoCommunicationError:
+                        rospy.logwarn(
+                            f'[{rospy.get_name()}] PantherCAN: SdoCommunicationError occurred while reading runtime status flag'
+                        )
+                        self._error_handle()
+                    yield motor_controller.runtime_stat_flag[i]
+        
+    def _error_handle(self) -> None:
+        self._err_times.append(time())
+        self._err_times.pop(0)
```

Comment:
If everything is lock guarded, let this function be as well. So this class will be fully thread safe

Reply:
This funcition is called only under lock. This will cause the program to stop.